File: Codes from the AI-assistants/Old Experiments/[CLBG] k-nucleotide/Basic Prompt - Normal Mode/KNucleotide-claude-basic-normal.py
#!/usr/bin/env python3
"""
K-nucleotide frequency analysis implementation
Reads FASTA format from stdin and analyzes DNA sequence THREE
"""

import sys
from collections import defaultdict
import operator

def read_fasta_sequence_three(file_handle):
    """Read FASTA format and extract DNA sequence THREE"""
    sequences = {}
    current_seq = None
    current_data = []
    
    for line in file_handle:
        line = line.strip()
        if line.startswith('>'):
            if current_seq is not None:
                sequences[current_seq] = ''.join(current_data)
            current_seq = line[1:]  # Remove '>' prefix
            current_data = []
        else:
            current_data.append(line.upper())
    
    # Don't forget the last sequence
    if current_seq is not None:
        sequences[current_seq] = ''.join(current_data)
    
    # Extract sequence THREE
    for seq_name, seq_data in sequences.items():
        if 'THREE' in seq_name.upper():
            return seq_data
    
    # If no sequence named THREE found, return the last/longest sequence
    if sequences:
        return max(sequences.values(), key=len)
    
    return ""

def update_hash_table(hash_table, dna_sequence, k):
    """
    Update hash table with k-nucleotide counts for all reading frames
    Args:
        hash_table: dictionary to store k-nucleotide counts
        dna_sequence: DNA sequence string
        k: length of nucleotide subsequences
    """
    seq_len = len(dna_sequence)
    
    # Process all possible k-nucleotides (all reading frames)
    for i in range(seq_len - k + 1):
        k_nucleotide = dna_sequence[i:i + k]
        # Use defaultdict to automatically initialize counts to 0
        hash_table[k_nucleotide] += 1

def calculate_frequencies(hash_table, total_count):
    """Calculate percentage frequencies and sort by descending frequency, then ascending key"""
    freq_list = []
    for nucleotide, count in hash_table.items():
        percentage = (count * 100.0) / total_count
        freq_list.append((percentage, nucleotide, count))
    
    # Sort by descending frequency, then ascending nucleotide key
    freq_list.sort(key=lambda x: (-x[0], x[1]))
    return freq_list

def print_frequency_table(hash_table, k):
    """Print frequency table sorted by descending frequency"""
    total_count = sum(hash_table.values())
    if total_count == 0:
        return
        
    freq_list = calculate_frequencies(hash_table, total_count)
    
    for percentage, nucleotide, count in freq_list:
        print(f"{nucleotide} {percentage:.3f}")
    print()

def print_specific_counts(hash_table, targets):
    """Print counts for specific nucleotide sequences"""
    for target in targets:
        count = hash_table.get(target, 0)
        print(f"{count}\t{target}")

def main():
    # Read DNA sequence THREE from stdin
    dna_sequence = read_fasta_sequence_three(sys.stdin)
    
    if not dna_sequence:
        print("No DNA sequence found", file=sys.stderr)
        return
    
    # Hash tables for different k-nucleotide lengths
    hash_tables = {}
    
    # Count 1-nucleotide and 2-nucleotide sequences
    for k in [1, 2]:
        hash_tables[k] = defaultdict(int)
        update_hash_table(hash_tables[k], dna_sequence, k)
        print_frequency_table(hash_tables[k], k)
    
    # Count 3, 4, 6, 12, and 18-nucleotide sequences for specific targets
    target_sequences = {
        3: ["GGT"],
        4: ["GGTA"],
        6: ["GGTATT"],
        12: ["GGTATTTTAATT"],
        18: ["GGTATTTTAATTTATAGT"]
    }
    
    for k in [3, 4, 6, 12, 18]:
        hash_tables[k] = defaultdict(int)
        update_hash_table(hash_tables[k], dna_sequence, k)
        
        if k in target_sequences:
            print_specific_counts(hash_tables[k], target_sequences[k])

if __name__ == "__main__":
    main()

File: Codes from the AI-assistants/Old Experiments/[CLBG] k-nucleotide/Basic Prompt - Normal Mode/KNucleotide-claude-basic-normal.java
import java.io.*;
import java.util.*;

/**
 * K-nucleotide frequency analysis implementation
 * Reads FASTA format from stdin and analyzes DNA sequence THREE
 */
public class KNucleotide {
    
    public static String readFastaSequenceThree(BufferedReader reader) throws IOException {
        Map<String, StringBuilder> sequences = new HashMap<>();
        String currentSeq = null;
        StringBuilder currentData = new StringBuilder();
        String line;
        
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (line.startsWith(">")) {
                if (currentSeq != null) {
                    sequences.put(currentSeq, new StringBuilder(currentData.toString()));
                }
                currentSeq = line.substring(1); // Remove '>' prefix
                currentData = new StringBuilder();
            } else {
                currentData.append(line.toUpperCase());
            }
        }
        
        // Don't forget the last sequence
        if (currentSeq != null) {
            sequences.put(currentSeq, currentData);
        }
        
        // Extract sequence THREE
        for (Map.Entry<String, StringBuilder> entry : sequences.entrySet()) {
            if (entry.getKey().toUpperCase().contains("THREE")) {
                return entry.getValue().toString();
            }
        }
        
        // If no sequence named THREE found, return the longest sequence
        if (!sequences.isEmpty()) {
            return sequences.values().stream()
                .max(Comparator.comparing(sb -> sb.length()))
                .map(StringBuilder::toString)
                .orElse("");
        }
        
        return "";
    }
    
    /**
     * Update hash table with k-nucleotide counts for all reading frames
     */
    public static void updateHashTable(Map<String, Integer> hashTable, String dnaSequence, int k) {
        int seqLen = dnaSequence.length();
        
        // Process all possible k-nucleotides (all reading frames)
        for (int i = 0; i <= seqLen - k; i++) {
            String kNucleotide = dnaSequence.substring(i, i + k);
            hashTable.put(kNucleotide, hashTable.getOrDefault(kNucleotide, 0) + 1);
        }
    }
    
    public static void printFrequencyTable(Map<String, Integer> hashTable) {
        int totalCount = hashTable.values().stream().mapToInt(Integer::intValue).sum();
        if (totalCount == 0) return;
        
        List<Map.Entry<String, Integer>> entries = new ArrayList<>(hashTable.entrySet());
        
        // Sort by descending frequency, then ascending nucleotide key
        entries.sort((a, b) -> {
            double freqA = (a.getValue() * 100.0) / totalCount;
            double freqB = (b.getValue() * 100.0) / totalCount;
            int freqCompare = Double.compare(freqB, freqA); // Descending frequency
            if (freqCompare != 0) return freqCompare;
            return a.getKey().compareTo(b.getKey()); // Ascending key
        });
        
        for (Map.Entry<String, Integer> entry : entries) {
            double percentage = (entry.getValue() * 100.0) / totalCount;
            System.out.printf("%s %.3f%n", entry.getKey(), percentage);
        }
        System.out.println();
    }
    
    public static void printSpecificCounts(Map<String, Integer> hashTable, String[] targets) {
        for (String target : targets) {
            int count = hashTable.getOrDefault(target, 0);
            System.out.printf("%d\t%s%n", count, target);
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        
        // Read DNA sequence THREE from stdin
        String dnaSequence = readFastaSequenceThree(reader);
        
        if (dnaSequence.isEmpty()) {
            System.err.println("No DNA sequence found");
            return;
        }
        
        // Hash tables for different k-nucleotide lengths
        Map<Integer, Map<String, Integer>> hashTables = new HashMap<>();
        
        // Count 1-nucleotide and 2-nucleotide sequences
        for (int k : new int[]{1, 2}) {
            hashTables.put(k, new HashMap<>());
            updateHashTable(hashTables.get(k), dnaSequence, k);
            printFrequencyTable(hashTables.get(k));
        }
        
        // Count 3, 4, 6, 12, and 18-nucleotide sequences for specific targets
        Map<Integer, String[]> targetSequences = new HashMap<>();
        targetSequences.put(3, new String[]{"GGT"});
        targetSequences.put(4, new String[]{"GGTA"});
        targetSequences.put(6, new String[]{"GGTATT"});
        targetSequences.put(12, new String[]{"GGTATTTTAATT"});
        targetSequences.put(18, new String[]{"GGTATTTTAATTTATAGT"});
        
        for (int k : new int[]{3, 4, 6, 12, 18}) {
            hashTables.put(k, new HashMap<>());
            updateHashTable(hashTables.get(k), dnaSequence, k);
            
            if (targetSequences.containsKey(k)) {
                printSpecificCounts(hashTables.get(k), targetSequences.get(k));
            }
        }
        
        reader.close();
    }
}